parens.go

```diff
@@ -1,88 +1,156 @@
-	// and return the expanded version. If given form is not macro form,
-	// it should return nil, nil.
-	Expand(ev Evaluator, form value.Any) (value.Any, error)
+// Context represents the environment/context in which forms are evaluated
+// for result. Context is not safe for concurrent use. Use fork() to get a
+// child context for concurrent executions.
+type Context struct {
+	ctx        context.Context
+	parent     *Context
+	analyzer   Analyzer
+	expander   Expander
+	stack      []stackFrame
+	maxDepth   int
+	mapFactory func() ConcurrentMap
 }
 
-// Eval performs macro-expansion if necessary, converts the expanded form to an
-// expression and evaluates the resulting expression.
-func (ev Evaluator) Eval(ctx Context, form value.Any) (value.Any, error) {
-	if form == nil {
-		return nil, nil
-	}
-
-	expr, err := ev.expandAnalyze(form)
+// Eval performs macro-expansion if necessary, converts the expanded form
+// to an expression and evaluates the resulting expression.
+func (ctx *Context) Eval(form value.Any) (value.Any, error) {
+	expr, err := ctx.expandAnalyze(form)
 	if err != nil {
 		return nil, err
 	} else if expr == nil {
 		return nil, nil
 	}
 
-	return expr.Eval(ctx, ev)
+	return expr.Eval(ctx)
 }
 
-func (ev Evaluator) expandAnalyze(form value.Any) (Expr, error) {
+func (ctx *Context) expandAnalyze(form value.Any) (Expr, error) {
 	if expr, ok := form.(Expr); ok {
 		// Already an Expr, nothing to do.
 		return expr, nil
 	}
 
-	if expanded, err := ev.expand(form); err != nil {
+	if expanded, err := ctx.expander.Expand(ctx, form); err != nil {
 		return nil, err
 	} else if expanded != nil {
 		// Expansion did happen. Throw away the old form and continue with
 		// the expanded version.
 		form = expanded
 	}
 
-	return ev.analyze(form)
+	return ctx.analyzer.Analyze(ctx, form)
 }
 
-func (ev Evaluator) expand(form value.Any) (value.Any, error) {
-	if ev.Expander == nil {
-		return basicExpander{}.Expand(ev, form)
+// fork creates a child context from the parent and returns, which can be
+// used as context for an independent thread of execution.
+func (ctx *Context) fork() *Context {
+	return &Context{
+		ctx:        ctx.ctx,
+		parent:     ctx,
+		stack:      append([]stackFrame(nil), ctx.stack...),
+		expander:   ctx.expander,
+		analyzer:   ctx.analyzer,
+		maxDepth:   ctx.maxDepth,
+		mapFactory: ctx.mapFactory,
 	}
+}
 
-	return ev.Expander.Expand(ev, form)
+func (ctx *Context) push(frame stackFrame) {
+	ctx.stack = append(ctx.stack, frame)
 }
 
-func (ev Evaluator) analyze(form value.Any) (Expr, error) {
-	if ev.Analyzer == nil {
-		return BasicAnalyzer{}.Analyze(ev, form)
+func (ctx *Context) pop() (frame *stackFrame) {
+	if len(ctx.stack) == 0 {
+		panic("pop from empty stack")
 	}
+	frame, ctx.stack = &ctx.stack[len(ctx.stack)-1], ctx.stack[:len(ctx.stack)-1]
+	return frame
+}
+
+func (ctx *Context) setGlobal(key string, value value.Any) {
+	// TODO: verify this. what is expected if this is a child context?
```

Comment:
This TODO is from me. Just to confirm we are on the same page, you are saying the current behaviour is correct. Current behaviour is basically:

1. Locate root context (if this the main thread, then `ctx` itself is the root. otherwise, locate the root by following `parent` pointer)
2. Do a `Store()` into the 0th frame of the stack in root context.

Reply:
>you are saying the current behaviour is correct.
Correct.

>1. Locate root context (if this the main thread, then ctx itself is the root. otherwise, locate the root by following parent pointer)
>2. Do a Store() into the 0th frame of the stack in root context.

Whoops, I had overlooked the fact that we are now storing a pointer to the parent `Context`.  Why are we doing this?
It seems like [we are still copying the full stack](https://github.com/spy16/parens/pull/20/files/bb7704e1c9191c3e1a10b345608c8c5cf2938344#diff-093c70a89dda8ea47ef4bc38df892a5bR86) from the parent to the child context.  Why chase pointers instead of using `ctx.stack[0].Store(...)`?

Reply:
actually the full copy is wrong.. but pointer chasing would be needed. 

1. Imagine you are currently in the root env `E1` and running some code in thread `T1`. You now launch `T2` which forks `E1` to create `E2`. You now have `T1` and `T2` running concurrently against `E1` and `E2` respectively. 
2. Let's say `T1` modifies a global binding in `E1`. This needs to be reflected in `T2` as well since its a global binding. 

If we took a copy of the stack during fork, this update by T1 would be lost to T2.

Reply:
>If we took a copy of the stack during fork, this update by T1 would be lost to T2.

Hmm, I don't think that's right.

Recall that [`stackFrame` holds a `ConcurrentMap` interface](https://github.com/spy16/parens/blob/feature/interpreter/parens.go#L155), which is a pointer type.  So while we are copying the stack, each stack frame contains a _pointer_ to a concurrent map (again, because interfaces are "just pointers").

https://play.golang.org/p/s9-7k04wl83

Reply:
I think there's a mistake in this design. Vars in a stack-frame cannot be shared across frames/envs except for frame-0 of root env. For example, consider that we are currently invoking `foo` so there is a corresponding frame in the stack with the local bindings (arguments, let vars etc.). Now `foo` invokes `bar`, so we push a frame and execute `bar`. But key requirement here is that we isolate the frame (i.e., vars from `foo` frame should not be leaked to `bar`).. 

With that in mind, I think we don't need a concurrent-map per frame. We just need a native map and only at Env level we need to maintain a single concurrent map. So we need to change the structure to this:

```go
type Env struct {
    globals ConcurrentMap
    stack []stackFrame
    // .... other fields ....
}
``` 

```go
type stackFrame struct {
    args []Any
    name string
    vars map[string]Any
}
```

(That also means, we don't need to push a `global` frame when we init into the stack. we just keep empty stack and the global frame is implied)

Reply:
Ah, ok.  I think I see what you mean!  My initial concern was that concurrent read/writes to local stack frames was possible, but I'm beginning to think this is not possible in Lisp.  Assuming that is indeed the case, then I'm 100% on board -- I would love nothing more than to get native map performance on local frames! 😃 

As a sanity check, and to make sure we're on the same page, consider this contrived Go code:

```go
func main() {
    var x int
    go print(x) // race condition:  may read x while x is incrementing
    x++
}
```

In this example, some sort of synchronization would be needed on `x`.  My concern was that a similar thing could occur in lisp code, where `foo` calls `(go bar)`, and then modifies a foo-local variable that `bar` is also reading/modifying.  If this were possible, we would need a `ConcurrentMap` for local stack frames as well.

Upon further reflection, it seems like this scenario is impossible Lisp due to how local variables are bound.  The reasoning is:

1. Assumption 1:  local variables can only be bound with `let`.
2. Assumption 2:  `let` pushes a new frame onto the stack.

Therefore:  once bound, a local variable cannot be rebound or modified.  

Is this correct?

Reply:
Addendum:  can we still have an option to set the `ConcurrentMap` implementation?  Map + `sync.RWMutex` is almost always the right choice, but I'd still like the possibility of using a more optimized datastructure if needed.

Reply:
More or less. But I don't think it's specific to how Lisp. It's usually how scopes work. Even in Go, you couldn't share the bindings map of a stack frame. Consider this example:

```go
func main() {
   foo := 100
   go someFunc()
   foo += 100
}

func someFunc() {
   // this is an error because the stack frame corresponding to invocation of `main`
   // and its variables map are not shared to the stack frame corresponding to this
   // function call (i.e., someFunc).
   fmt.Println(foo) 
}
```

Reply:
Ah good point -- I was confusing closures with function calls. 👍 

Thanks.  Resolving this now.